src/ComfortPay/Message/ChargeRequest.php

```diff
@@ -18,79 +24,163 @@ public function setTerminalId($value)
 
         $data = parent::getData();
         $data = array_merge($data, [
-            'terminalId' => $this->getTerminalId(),
-            'ws' => $this->getWs(),
-            'cid' => $this->getCid(),
+            'transactionType' => $this->getTransactionType(),
             'transactionId' => $this->getTransactionId(),
+            'parentTransactionId' => $this->getParentTransactionId(),
+            'referedCardId' => $this->getReferedCardId(),
+            'merchantId' => $this->getMid(),
+            'terminalId' => $this->getTerminalId(),
             'amount' => $this->getAmount(),
-            'currency' => (new ISOCurrencies())->numericCodeFor(new Currency($this->getCurrency())),
+            'cc' => $this->getCurrency(),
+            'ipspData' => $this->getIpspData(),
             'vs' => $this->getVs(),
             'ss' => $this->getSs(),
+            'e2eReference' => $this->getE2eReference(),
+            'submerchantId' => $this->getSubmerchantId(),
+            'location' => $this->getLocation(),
+            'city' => $this->getCity(),
+            'alpha2CountryCode' => $this->getAlpha2CountryCode(),
         ]);
-        
+
         return $data;
     }
 
     public function sendData($data)
     {
-        if ($this->getTestmode()) {
+        if ($this->getTestmode())
+        {
             if (intval($data['cid']) % 2 == 0) {
-                return $this->response = new ChargeResponse($this, ['transactionStatus' => '02', 'transactionApproval' => '123']);
+                return $this->response = new ChargeResponse($this,
+                    ['transactionId' => $data['transactionId'], 'transactionStatus' => '02', 'transactionApproval' => '123']);
             }
-            return $this->response = new ChargeResponse($this, ['transactionStatus' => '00', 'transactionApproval' => '123']);
+            return $this->response = new ChargeResponse($this,
+                ['transactionId' => $data['transactionId'], 'transactionStatus' => '00', 'transactionApproval' => '123']);
+        }
+
+        $req = new \stdClass();
+        $req->transactionId = $data['transactionId'];
+        $req->referedCardId = $data['referedCardId'];
+        $req->merchantId = $data['merchantId'];
+        $req->terminalId = $data['terminalId'];
+        $req->amount = $data['amount'];
+        $req->parentTransactionId = $data['parentTransactionId'];
+        $req->cc = $data['cc'];
+
+        $symbols = new \stdClass();
+        $symbols->variableSymbol = $data['vs'];
+        $symbols->specificSymbol = $data['ss'];
+
+        $transactionIdentificator = new \stdClass();
+        $transactionIdentificator->symbols = $symbols;
+        $transactionIdentificator->e2eReference = $data['e2eReference'];
+        $req->transactionIdentificator = $transactionIdentificator;
+
+        if (!empty($data['submerchantId']) && !empty($data['location']) && !empty($data['city']) && !empty($data['alpha2CountryCode'])) {
```

Comment:
This is not compliant with the note in the WSDL:

> `ipspData`, sice nepovinne, ale ak sa v requete uvedie, MUSIA byt vnutorne fieldy vyplnene

That means, that if at least one of them is provided and others are missing, you should return validation error from the library stating that the rest is missing.

Right now if client provides 3 out of 4 values, you ignore them silently and client has no idea something is wrong.